### PR TITLE
Element Metric Fixes and Extensions

### PR DESCRIPTION
**Pull request by rinkk**
this PR addresses a number of things concerned with mesh element metrics. I got a bit carried away and now it's all mixed up a bit. sorry!

in detail this changed the following:
* i cleaned up and restructured code for vis-pipeline, i.e. removed algorithmic code that should not be in vis-pipeline into a new element metric interface class, put histogram code into histogram class, etc.
* histograms for element metrics can now be output in arbitrary files (build into gui, but available from interface class in meshlib)
* area- and volume metrics are now combined into one "size metric" that can now also handle 1d meshes
* range values for the new size metric are no longer normalised to [0,1] but are within their "real" range, i.e. you can select all element with ```size < x``` and things like that
* a new quality metric has been added based on size differences between connected elements (i.e. large differences in size are bad, roughly similar size is good)

### CODE REVIEW

File: MeshLib/MeshQuality/SizeDifferenceMetric.h [hunk 1]
```diff
@@ -23,14 +23,14 @@ namespace MeshLib
 /** 
  * Calculates the quality of mesh elements based on the volume of elements
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/SizeDifferenceMetric.h]: please update the description

File: Applications/DataExplorer/VtkVis/VtkCompositeElementSelectionFilter.h [hunk 2]
```diff
@@ -32,6 +32,9 @@ class VtkCompositeElementSelectionFilter : public VtkCompositeFilter
 
 	virtual void init();
 
+	// Sets the range for the quality measure (default is [0,1] but this may vary for area- and volume-metrics).
+	void setRange(double min_val, double max_val) { _range = std::pair<double,double>(min_val, max_val); }
```
Comment (endJunction) [inline Applications/DataExplorer/VtkVis/VtkCompositeElementSelectionFilter.h]: A shorter version is:
```c++
_range = std::make_pair(min_val, max_val);
```

**Comment by endJunction** [comment]
:fast_forward:

File: MeshLib/MeshQuality/ElementQualityInterface.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+	/// Returns a histogram of the quality vector seperated into the given number of bins.
+	/// If no number of bins is specified, one will be calculated based on the Sturges criterium.
+	BaseLib::Histogram<double> getHistogram(std::size_t n_bins = 0) const
+	{
+		return _quality_tester->getHistogram(static_cast<size_t>(n_bins));
+	}
+
+	/// Writes a histogram of the quality vector to a specified file.
+	int writeHistogram(std::string const& file_name, std::size_t n_bins = 0) const
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementQualityInterface.h]: Is this function really needed? A user can write like

```
ElementQualityInterface obj;
obj.getHistogram().write(...);
```

File: MeshLib/MeshQuality/ElementQualityInterface.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	/// Destructor
+	~ElementQualityInterface() 
+	{ 
+		delete _quality_tester; 
+	}
+
+	/// Returns the vector containing a quality measure for each element.
+	std::vector<double> const& getQualityVector() const
+	{ 
+		return _quality_tester->getElementQuality();
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementQualityInterface.h]: need to check quality_tester!=nullptr

File: MeshLib/MeshQuality/ElementQualityInterface.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	std::vector<double> const& getQualityVector() const
+	{ 
+		return _quality_tester->getElementQuality(); 
+	}
+
+	/// Returns a histogram of the quality vector seperated into the given number of bins.
+	/// If no number of bins is specified, one will be calculated based on the Sturges criterium.
+	BaseLib::Histogram<double> getHistogram(std::size_t n_bins = 0) const
+	{
+		return _quality_tester->getHistogram(static_cast<size_t>(n_bins));
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementQualityInterface.h]: need to check quality_tester!=nullptr

File: MeshLib/MeshQuality/ElementQualityInterface.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+			_quality_tester = new MeshLib::ElementSizeMetric(mesh);
+		else if (t == MeshQualityType::SIZEDIFFERENCE)
+			_quality_tester = new MeshLib::SizeDifferenceMetric(mesh);
+		else if (t == MeshQualityType::EQUIANGLESKEW)
+			_quality_tester = new MeshLib::AngleSkewMetric(mesh);
+		else if (t == MeshQualityType::RADIUSEDGERATIO)
+			_quality_tester = new MeshLib::RadiusEdgeRatioMetric(mesh);
+		else
+		{
+			ERR("VtkVisPipeline::checkMeshQuality(): Unknown MeshQualityType.");
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementQualityInterface.h]: The function name in the error message is incorrect

File: MeshLib/MeshQuality/ElementSizeMetric.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
+/**
+ * \file   ElementSizeMetric.cpp
+ * \author Karsten Rink
+ * \date   2011-03-17
+ * \brief  Implementation of the AreaMetric class.
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementSizeMetric.cpp]: AreaMetric -> ElementSizeMetric

File: MeshLib/MeshQuality/ElementSizeMetric.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	const std::vector<MeshLib::Element*>& elements(_mesh.getElements());
+	const std::size_t nElems(elements.size());
+	std::size_t error_count(0);
+	
+	for (size_t k(0); k < nElems; k++)
+	{
+		Element const& elem (*elements[k]);
+		if (elem.getDimension()<3)
+		{
+            _element_quality_metric[k] = 0.0;
```
Comment (norihiro-w) [inline MeshLib/MeshQuality/ElementSizeMetric.cpp]: indentation

File: MeshLib/MeshQuality/ElementQualityMetric.cpp [hunk 9]
```diff
@@ -28,14 +28,12 @@ ElementQualityMetric::ElementQualityMetric(Mesh const& mesh) :
...
 }
 
-BaseLib::Histogram<double> ElementQualityMetric::getHistogram (size_t nclasses) const
+BaseLib::Histogram<double> ElementQualityMetric::getHistogram (size_t n_bins) const
 {
-	if (nclasses == 0) {
-		// simple suggestion: number of classes with Sturges criterion
-		nclasses = static_cast<size_t>(1 + 3.3 * log (static_cast<float>((_mesh.getNElements()))));
-	}
+	if (n_bins == 0)
```
Comment (TomFischer) [inline MeshLib/MeshQuality/ElementQualityMetric.cpp]: Please keep the comment in line 34.

File: MeshLib/MeshQuality/ElementSizeMetric.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	return error_count;
+}
+
+std::size_t ElementSizeMetric::calc3dQuality()
+{
+	const std::vector<MeshLib::Element*>& elements(_mesh.getElements());
+	const std::size_t nElems(elements.size());
+	std::size_t error_count(0);
+	
+	for (size_t k(0); k < nElems; k++)
```
Comment (TomFischer) [inline MeshLib/MeshQuality/ElementSizeMetric.cpp]: `std::size_t`

File: MeshLib/MeshQuality/SizeDifferenceMetric.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+#ifndef SIZEDIFFERENCEMETRIC_H_
+#define SIZEDIFFERENCEMETRIC_H_
+
+#include "ElementQualityMetric.h"
+
+namespace MeshLib
+{
+
+/** 
+ * Calculates the quality of mesh elements based the difference of element size in comparison to
```
Comment (TomFischer) [inline MeshLib/MeshQuality/SizeDifferenceMetric.h]: Is there an "on" missing (... based on ..)?

File: MeshLib/MeshQuality/ElementQualityInterface.h (hunk 3 above)
Comment (rinkk) [inline MeshLib/MeshQuality/ElementQualityInterface.h]: I am not sure. I think I'd prefer to keep this method here to really have one interface that handles everything that is needed.

File: MeshLib/MeshQuality/ElementQualityMetric.cpp (hunk 9 above)
Comment (rinkk) [inline MeshLib/MeshQuality/ElementQualityMetric.cpp]: the comment has been moved into the header file to the documentation of this method.

**Comment by TomFischer** [comment]
:+1:

**Comment by norihiro-w** [comment]
:+1: besides a minor comment

**Pull request merged**